numba/targets/arrayobj.py

```diff
@@ -905,6 +905,44 @@ def array_argmax_impl(arry):
     return context.compile_internal(builder, array_argmax_impl, sig, args)
 
 
+@builtin
+@implement(numpy.median, types.Kind(types.Array))
+@implement("array.mean", types.Kind(types.Array))
+def array_median(context, builder, sig, args):
+    # XXX: the implementation doesn't work for the case if array is not one
+    # dimentional
+    def partition(A, p, r):
+        # TODO: make it randomized
+        x = A[r]
+        i = p-1
+        for j in range(p, r):
+            if A[j] <= x:
+                i += 1
+                A[i], A[j] = A[j], A[i]
+        A[i+1], A[r] = A[r], A[i+1]
+        return i + 1
+
+    def select(arry, k):
+        n = arry.shape[0]
+        temp_arry = arry.copy()
+        i = partition(temp_arry, 0, n-1)
+        while i != k:
+            if i < k:
+                i = partition(temp_arry, i+1, n-1)
+            else:
+                i = partition(temp_arry, 0, i-1)
```

Comment:
If this is not the first iteration, it seems this is doing too much here.
For example: n = 15, k = 7.
- first partition call: i = partition(0, 14) -> 3
- second partition call: i = partition(4, 14) -> 12
- third partition call: partition(0, 11)?? you should be calling partition(4, 11)...